File: packages/soulseek_protocol/lib/src/soulseek_client.dart
import 'dart:async';

import 'connection/server_connection.dart';
import 'connection/socket_manager.dart';
import 'messages/buffer.dart';
import 'messages/codes.dart';
import 'messages/message.dart';
import 'peer/peer_connection.dart';
import 'services/auth_service.dart';
import 'services/browse_service.dart';
import 'services/chat_service.dart';
import 'services/search_service.dart';
import 'services/user_service.dart';
import 'services/wishlist_service.dart';
import 'transfer/download_manager.dart';

class SoulseekClient {
  final ServerTransport _server;
  final AuthService auth;
  final SearchService searchService;
  final ChatService chat;
  final UserService users;
  final BrowseService browseService;
  final WishlistService wishlistService;
  final DownloadManager _downloadManager;
  final Map<String, PeerConnection> _activePeers = {};

  SoulseekClient({
    ServerTransport? server,
    AuthService? auth,
    SearchService? searchService,
    ChatService? chat,
    UserService? users,
    BrowseService? browseService,
    WishlistService? wishlistService,
  }) : _server = server ?? ServerConnection(),
       auth = auth ?? AuthService(server: server ?? ServerConnection()),
       searchService = searchService ?? SearchService(server: server ?? ServerConnection()),
       chat = chat ?? ChatService(server: server ?? ServerConnection()),
       users = users ?? UserService(server: server ?? ServerConnection()),
       browseService = browseService ?? BrowseService(),
       wishlistService = wishlistService ?? WishlistService(server: server ?? ServerConnection()),
       _downloadManager = DownloadManager();

  Stream<ServerConnectionState> get connectionState => auth.connectionState;
  Stream<SearchResult> get searchResults => searchService.results;
  Stream<PrivateMessage> get privateMessages => chat.privateMessages;
  Stream<UserStatus> get userStatus => users.userStatus;
  Stream<Map<String, DownloadProgress>> get downloadProgress => _downloadManager.allProgress;
  Stream<WishlistReply> get wishlistResults => wishlistService.wishlistResults;
  Stream<ConnectionInfo> get connectionInfo => auth.connectionInfo;
  bool get authenticated => auth.authenticated;
  String? get username => auth.username;
  DownloadManager get downloadManager => _downloadManager;

  void init() {
    auth.init();
    searchService.init();
    chat.init();
    users.init();
    wishlistService.init();
  }

  Future<void> connect(String username, String password) async {
    await auth.connect(username, password);
  }

  Future<void> disconnect() async {
    await auth.disconnect();
    for (final peer in _activePeers.values) {
      await peer.disconnect();
    }
    _activePeers.clear();
  }

  void setListenPort(int port) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.setListenPort,
      SetListenPort(port).serialize().toBytes(),
    ));
  }

  int search(String query) => searchService.search(query);

  void addUser(String username) => users.addUser(username);

  void getPeerAddress(String username) => users.getPeerAddress(username);

  void sendPrivateMessage(String username, String message) {
    chat.sendPrivateMessage(username, message);
  }

  void checkDownloadQueue(List<String> usernames) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.checkDownloadQueue,
      CheckDownloadQueue(usernames).serialize().toBytes(),
    ));
  }

  Future<PeerConnection?> connectToPeer(String username, int ip, int port) async {
    if (_activePeers.containsKey(username)) {
      await _activePeers[username]!.disconnect();
      _activePeers.remove(username);
    }

    final socketManager = SocketManager();
    final peer = PeerConnection(
      username: username,
      ip: ip,
      port: port,
      type: PeerConnectionType.outgoing,
      socketManager: socketManager,
    );

    await peer.connect();
    _activePeers[username] = peer;
    return peer;
  }

  void enqueueDownload({
    required String filename,
    required int size,
    required String username,
    required int fileCode,
    String? localPath,
  }) {
    _downloadManager.addDownload(
      filename: filename,
      size: size,
      username: username,
      fileCode: fileCode,
      localPath: localPath,
    );
  }

  Future<void> requestDownload({
    required String filename,
    required int size,
    required String username,
    required int fileCode,
    required PeerConnection connection,
  }) async {
    final download = _downloadManager.addDownload(
      filename: filename,
      size: size,
      username: username,
      fileCode: fileCode,
    );

    await _downloadManager.startDownload(download, connection);
  }

  Future<FolderContentsReply> browseUser({
    required String username,
    required int ip,
    required int port,
    String directory = '',
  }) async {
    final connection = await connectToPeer(username, ip, port);
    if (connection == null) {
      throw Exception('Failed to connect to peer $username');
    }
    return browseService.browseUser(
      connection: connection,
      directory: directory,
    );
  }

  int wishlistSearch(String query) => wishlistService.wishlistSearch(query);

  void addWishlistItem(String phrase) => wishlistService.addWishlistItem(phrase);

  void removeWishlistItem(String phrase) => wishlistService.removeWishlistItem(phrase);

  void dispose() {
    searchService.dispose();
    chat.dispose();
    users.dispose();
    browseService.dispose();
    wishlistService.dispose();
    auth.dispose();
    _downloadManager.dispose();
    for (final peer in _activePeers.values) {
      peer.dispose();
    }
    _activePeers.clear();
  }
}

File: Soulseek.Protocol/SoulseekClient.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;
using Soulseek.Protocol.Peer;
using Soulseek.Protocol.Services;
using Soulseek.Protocol.Transfer;

namespace Soulseek.Protocol;

public class SoulseekClient
{
    private readonly IServerTransport _server;
    private readonly DownloadManager _downloadManager;
    private readonly UploadManager _uploadManager;
    private readonly Dictionary<string, PeerConnection> _activePeers = new();
    private PeerListener? _peerListener;

    public AuthService Auth { get; }
    public SearchService SearchService { get; }
    public ChatService Chat { get; }
    public UserService Users { get; }
    public BrowseService BrowseService { get; }
    public WishlistService WishlistService { get; }
    public RoomChatService RoomChat { get; }
    public UploadManager UploadManager => _uploadManager;
    public DownloadManager DownloadManager => _downloadManager;

    public SoulseekClient(
        IServerTransport? server = null,
        AuthService? auth = null,
        SearchService? searchService = null,
        ChatService? chat = null,
        UserService? users = null,
        BrowseService? browseService = null,
        WishlistService? wishlistService = null,
        RoomChatService? roomChat = null)
    {
        _server = server ?? new ServerConnection();
        Auth = auth ?? new AuthService(_server);
        SearchService = searchService ?? new SearchService(_server);
        Chat = chat ?? new ChatService(_server);
        Users = users ?? new UserService(_server);
        BrowseService = browseService ?? new BrowseService();
        WishlistService = wishlistService ?? new WishlistService(_server);
        RoomChat = roomChat ?? new RoomChatService(_server);
        _downloadManager = new DownloadManager();
        _uploadManager = new UploadManager();
    }

    public IObservable<ServerConnectionState> ConnectionState => Auth.ConnectionState;
    public IObservable<SearchResult> SearchResults => SearchService.Results;
    public IObservable<PrivateMessage> PrivateMessages => Chat.PrivateMessages;
    public IObservable<UserStatusMessage> UserStatus => Users.UserStatus;
    public IObservable<Dictionary<string, DownloadProgress>> DownloadProgress => _downloadManager.AllProgress;
    public IObservable<WishlistReply> WishlistResults => WishlistService.WishlistResults;
    public IObservable<RoomMessageData> RoomMessages => RoomChat.RoomMessages;
    public IObservable<UserJoinedRoom> RoomUserJoined => RoomChat.UserJoined;
    public IObservable<UserLeftRoom> RoomUserLeft => RoomChat.UserLeft;
    public IObservable<RoomList> RoomList => RoomChat.RoomList;
    public IObservable<ConnectionInfo> ConnectionInfo => Auth.ConnectionInfo;
    public IObservable<PeerAddressResponse> PeerAddressResponses => Users.PeerAddressResponses;
    public bool Authenticated => Auth.Authenticated;
    public string? Username => Auth.Username;

    public void Init()
    {
        Auth.Init();
        SearchService.Init();
        Chat.Init();
        Users.Init();
        WishlistService.Init();
        RoomChat.Init();
    }

    public Task Connect(string username, string password) => Auth.Connect(username, password);

    public async Task Disconnect()
    {
        await StopListening();
        await Auth.Disconnect();
        foreach (var peer in _activePeers.Values)
            await peer.Disconnect();
        _activePeers.Clear();
    }

    public void SetServer(string host, int port)
    {
        _server.SetServer(host, port);
    }

    public void SetListenPort(int port)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.SetListenPort,
            new SetListenPort(port).Serialize().ToBytes()
        ));
    }

    public async Task StartListening(int port)
    {
        if (_peerListener != null)
            await _peerListener.Stop();
        _peerListener = new PeerListener(_uploadManager);
        await _peerListener.Start(port);
        SetListenPort(port);
    }

    public async Task StopListening()
    {
        if (_peerListener != null)
        {
            await _peerListener.Stop();
            _peerListener = null;
        }
    }

    public int Search(string query) => SearchService.Search(query);
    public void AddUser(string username) => Users.AddUser(username);
    public void GetPeerAddress(string username) => Users.GetPeerAddress(username);

    public void SendPrivateMessage(string username, string message)
    {
        Chat.SendPrivateMessage(username, message);
    }

    public void CheckDownloadQueue(List<string> usernames)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.CheckDownloadQueue,
            new CheckDownloadQueue(usernames).Serialize().ToBytes()
        ));
    }

    public async Task<PeerConnection?> ConnectToPeer(string username, int ip, int port)
    {
        if (_activePeers.TryGetValue(username, out var existing))
        {
            await existing.Disconnect();
            _activePeers.Remove(username);
        }

        var socketManager = new SocketManager();
        var peer = new PeerConnection(username, ip, port, PeerConnectionType.Outgoing, socketManager);

        await peer.Connect();
        _activePeers[username] = peer;
        return peer;
    }

    public void EnqueueDownload(string filename, int size, string username, int fileCode, string? localPath = null)
    {
        _downloadManager.AddDownload(filename, size, username, fileCode, localPath);
    }

    public async Task RequestDownload(string filename, int size, string username, int fileCode, PeerConnection connection, string? localPath = null)
    {
        var download = _downloadManager.AddDownload(filename, size, username, fileCode, localPath);
        await _downloadManager.StartDownload(download, connection);
    }

    public async Task<FolderContentsReply> BrowseUser(string username, int ip, int port, string directory = "")
    {
        var connection = await ConnectToPeer(username, ip, port);
        if (connection == null)
            throw new Exception($"Failed to connect to peer {username}");
        return await BrowseService.BrowseUser(connection, directory);
    }

    public int WishlistSearch(string query) => WishlistService.WishlistSearch(query);
    public void AddWishlistItem(string phrase) => WishlistService.AddWishlistItem(phrase);
    public void RemoveWishlistItem(string phrase) => WishlistService.RemoveWishlistItem(phrase);
    public void JoinRoom(string roomName) => RoomChat.JoinRoom(roomName);
    public void LeaveRoom(string roomName) => RoomChat.LeaveRoom(roomName);
    public void SendRoomMessage(string roomName, string message) => RoomChat.SendMessage(roomName, message);
    public void RequestRoomList() => RoomChat.RequestRoomList();
    public void SetRoomTicker(string roomName, string ticker) => RoomChat.SetRoomTicker(roomName, ticker);
    public void RemoveRoomTicker(string roomName) => RoomChat.RemoveRoomTicker(roomName);

    public void Dispose()
    {
        SearchService.Dispose();
        Chat.Dispose();
        Users.Dispose();
        BrowseService.Dispose();
        WishlistService.Dispose();
        RoomChat.Dispose();
        Auth.Dispose();
        _peerListener?.Dispose();
        _downloadManager.Dispose();
        _uploadManager.Dispose();
        foreach (var peer in _activePeers.Values)
            peer.Dispose();
        _activePeers.Clear();
    }
}